docs/auth/backend-auth.md

```diff
@@ -0,0 +1,82 @@
+---
+id: backend-auth
+title: Secure your backend
+description: Documentation on securing your backend by requiring authentication.
+---
+
+##
+
+By default, the backstage backend does not require authentication. Securing your backend requires additional configuration or a separate service like Google's Identity-Aware Proxy. Since the backend is served as an express app we can leverage express middleware to require JWT bearer tokens for all requests. The `@backstage/plugin-auth-node` package provides several factory methods for creating express middleware that can be used to secure your backend:
+
+- `createRequireJwtTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests.
+- `createRequireUserTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests and verifies that the token is a valid user token.
+- `createRequireServerTokenMiddleware`: Creates middleware a JWT bearer token for all requests and verifies that the token is a valid server token (see [service to service Auth](./service-to-service-auth.md)).
+
+> NOTE: `createRequireJwtTokenMiddleware` does not validate the JWT tokens, instead you must specify an express middleware that will be invoked when the token has the specified audience. `createRequireUserTokenMiddleware` and `createRequireServerTokenMiddleware` both validate the tokens.
+
+The middleware created by these factories only verifies that a valid JWT token is provided and thus adds **authentication** to the api endpoint. The middleware cannot check whether the identity enclosed in the token has access to specific resources (**authorisation**). Each plugin backend is responsible for checking the permissions of the identity in the token if required.
+
+To configure all backend routes, except `/auth`, to require a valid user or server token, you can use the following code:
+
+```ts title="packages/backend/src/index.ts"
+// ...
+/* highlight-add-start */
+import {
+  createRequireJwtTokenMiddleware,
+  createRequireServerTokenMiddleware,
+  createRequireUserTokenMiddleware,
+} from '@internal/plugin-auth-node';
+/* highlight-add-end */
+
+// ...
+async function main() {
+  // ...
+  const createEnv = makeCreateEnv(config);
+
+  /* highlight-add-start */
+  const tokenManager = ServerTokenManager.fromConfig(config, {
+    logger: getRootLogger(),
+  });
+  const discovery = SingleHostDiscovery.fromConfig(config);
+  const identity = DefaultIdentityClient.create({
+    discovery,
+  });
+
+  const requireBackstageToken = createRequireJwtTokenMiddleware({
+    backstage: createRequireUserTokenMiddleware(identity),
+    'backstage-server': createRequireServerTokenMiddleware(tokenManager),
+  });
```

Comment:
How come you made this part of the user setup? I'm thinking the user/server token verification logic is something we can leave as internal, basically only having a single middleware.

Reply:
My initial thinking was that you can opt-in to either user tokens, server tokens, or both for your routes. But I agree that exposing a single middleware is a better option.

Reply:
Also this middleware is a more of a decorator. It uses the audience of the token to decide which middleware (user or server) to invoke to validate the token. However I found that the server token does not have an audience set so you can't use that to pick the right middleware anyway.

Reply:
I removed this and instead expose `createRequireAuthenticationMiddleware`